openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route_test.go

```diff
@@ -0,0 +1,131 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+// TestValidateRoutePlugin verifies if the route validation plugin can handle admits
+// for the resource {group: api/route/v1, kind: Route}
+func TestValidateRoutePlugin(t *testing.T) {
+	plugin, err := NewValidateRoute()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	routeAdmitter := plugin.(*validateCustomResourceWithClient)
+	routeAdmitter.setupWithFakeClient()
+
+	validator, ok := plugin.(admission.ValidationInterface)
+	if !ok {
+		t.Fatalf("could not construct validator from route validation plugin")
+	}
+
+	r1 := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "foo",
+			Namespace: "bar",
+		},
+		Spec: routev1.RouteSpec{
+			To: routev1.RouteTargetReference{
+				Kind: "Service",
+				Name: "default",
+			},
+		},
+	}
+	r2 := r1.DeepCopy()
+
+	s1 := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "foo",
+			Namespace: "bar",
+		},
+		Data: map[string][]byte{},
+	}
+	s2 := s1.DeepCopy()
+
+	testCases := []struct {
+		description string
+
+		object    runtime.Object
+		oldObject runtime.Object
+
+		kind     schema.GroupVersionKind
+		resource schema.GroupVersionResource
+
+		name      string
+		namespace string
+
+		expectedError bool
+	}{
+		{
+			description: "route object is passed to admission plugin with scheme routev1.Route",
+
+			object:    runtime.Object(r1),
+			oldObject: runtime.Object(r2),
+
+			kind:     routev1.GroupVersion.WithKind("Route"),
+			resource: routev1.GroupVersion.WithResource("routes"),
+
+			name:      r1.Name,
+			namespace: r1.Namespace,
+
+			expectedError: false,
+		},
+		{
+			description: "non-route object is passed to admission plugin with scheme corev1.Secret",
+
+			object:    runtime.Object(s1),
+			oldObject: runtime.Object(s2),
+
+			kind:     corev1.SchemeGroupVersion.WithKind("Secret"),
+			resource: corev1.SchemeGroupVersion.WithResource("secrets"),
+
+			name:      s1.Name,
+			namespace: s1.Namespace,
+
+			expectedError: false,
+		},
+		{
+			description: "non-route object is passed to admission plugin with conflicting scheme routev1.Route",
+
+			object:    runtime.Object(s1),
+			oldObject: runtime.Object(s2),
+
+			kind:     routev1.GroupVersion.WithKind("Route"),
+			resource: routev1.GroupVersion.WithResource("routes"),
+
+			name:      s1.Name,
+			namespace: s1.Namespace,
+
+			expectedError: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+
+			attr := admission.NewAttributesRecord(
+				tc.object, tc.oldObject,
+				tc.kind, tc.name, tc.namespace, tc.resource,
+				"", admission.Create, nil, false,
+				&user.DefaultInfo{},
+			)
+
+			err = validator.Validate(context.Background(), attr, nil)
+			if err != nil {
+				if !tc.expectedError {
+					t.Fatalf("admission error not expected, but found %q", err)
+				}
+			}
```

Comment:
```suggestion
			switch err = validator.Validate(context.Background(), attr, nil); {
			case !tc.expectedError && err != nil:
				t.Fatalf("admission error not expected, but found %q", err)
			case tc.expectedError && err == nil:
				t.Fatal("admission error expected, but got nil")
			}
```

Reply:
applied in https://github.com/openshift/kubernetes/commit/9f0d0d112aca0abeb1428b6828ea994be7afad6d.